Issue #813: Clear twig caches on deployments.

danepowell opened the pull request:
Fixes #813 

Changes proposed:
- Add `deploy:id` step to deployments that writes current deploy tag (if present) to a file in the deploy dir.
- Modify settings.php to read this new tag from disk (if it exists) and set `$settings['deployment_identifier']` accordingly.
- Patch core to use `$settings['deployment_identifier']` to handle stale Twig caches on multiple web heads.

I've tested this on a real project and confirmed that the identifier is set correctly, although I haven't been able to test that this actually fixes the problem with stale Twig templates, since that involves a production deploy. I'm fairly confident it will work though.

grasmash commented on settings/blt.settings.php:
```diff
@@ -80,6 +80,17 @@
+ * custom code that changes the container, changing this identifier will also
+ * allow the container to be invalidated as soon as code is deployed.
+ */
+$deploy_id = file_get_contents(DRUPAL_ROOT . '/../deploy_id.txt');
```
I think we should use a file_exists() check here.

grasmash commented on phing/tasks/deploy.xml:
```diff
@@ -139,6 +139,16 @@
+      <isset property="deploy.tag"/>
+      <then>
+        <!-- Write tag name for the benefit of environments like ACE that can't access it. -->
+        <exec dir="${deploy.dir}" command="echo '${deploy.tag}' > deploy_id.txt" logoutput="true" checkreturn="true" level="${blt.exec_level}"/>
```
Why not use the commit hash so that this works for non-tags as well? E.g., `git rev-parse HEAD` rather than `echo '${deploy.tag}'`.

grasmash replied:
I suggest that we name this file `deployment_identifier` rather than `deploy_id.txt` for consistency.

grasmash commented on phing/tasks/deploy.xml:
```diff
@@ -139,6 +139,16 @@
 
+  <target name="deploy:id" description="Writes deployment id information to artifact.">
+    <if>
+      <isset property="deploy.tag"/>
```
Why only do this for tags? I seems like we'd want it for all build artifacts.

danepowell commented:
I renamed the file to deployment_identifier and added a file_exists to catch notices.

The only time you'd need to set this deployment identifier is for production deploys, since dev/stage don't have multiple web heads. And you'd only ever deploy tags to production. So this seemed like a simpler solution than commit hashes. I also thought commit hashes would be weird because the deployment_identifier would be the _previous_ commit hash; it wouldn't match the current commit because writing the identifier obviously happens prior to the commit. I don't see why that would be bad, it's just weird.

Let me know if you'd rather use the commit hash, that's fine with me.

grasmash commented:
Makes sense.
